packages/integration-tests/src/test-project-setup/reference_auth_project.ts

```diff
@@ -0,0 +1,329 @@
+import { TestProjectBase } from './test_project_base.js';
+import fsp from 'fs/promises';
+import { createEmptyAmplifyProject } from './create_empty_amplify_project.js';
+import { CloudFormationClient } from '@aws-sdk/client-cloudformation';
+import { TestProjectCreator } from './test_project_creator.js';
+import { AmplifyClient } from '@aws-sdk/client-amplify';
+import { AuthResourceCreator } from '../resource-creation/auth_resource_creator.js';
+import { CognitoIdentityProviderClient } from '@aws-sdk/client-cognito-identity-provider';
+import { CognitoIdentityClient } from '@aws-sdk/client-cognito-identity';
+import { IAMClient } from '@aws-sdk/client-iam';
+import { BackendIdentifier } from '@aws-amplify/plugin-types';
+
+/**
+ * Creates a reference auth project
+ */
+export class ReferenceAuthTestProjectCreator implements TestProjectCreator {
+  readonly name = 'reference-auth';
+
+  /**
+   * Creates project creator.
+   */
+  constructor(
+    private readonly cfnClient: CloudFormationClient,
+    private readonly amplifyClient: AmplifyClient,
+    private readonly cognitoIdentityProviderClient: CognitoIdentityProviderClient,
+    private readonly cognitoIdentityClient: CognitoIdentityClient,
+    private readonly iamClient: IAMClient
+  ) {}
+
+  createProject = async (e2eProjectDir: string): Promise<TestProjectBase> => {
+    const { projectName, projectRoot, projectAmplifyDir } =
+      await createEmptyAmplifyProject(this.name, e2eProjectDir);
+
+    const project = new ReferenceAuthTestProject(
+      projectName,
+      projectRoot,
+      projectAmplifyDir,
+      this.cfnClient,
+      this.amplifyClient,
+      this.cognitoIdentityProviderClient,
+      this.cognitoIdentityClient,
+      this.iamClient
+    );
+
+    await fsp.cp(
+      project.sourceProjectAmplifyDirURL,
+      project.projectAmplifyDirPath,
+      {
+        recursive: true,
+      }
+    );
+
+    // generate resources
+    const {
+      userPool,
+      userPoolClient,
+      identityPool,
+      authRole,
+      unauthRole,
+      adminGroup,
+    } = await project.setupTestResources();
```

Comment:
Doing this before `deploy` is good idea.

But to make it fit into existing abstraction better we should pursue one of these options:
1. Override `deploy` to add extra processing. Like here https://github.com/aws-amplify/amplify-backend/blob/22aa7a659168c6e04a39343e4a52b712b6372e9c/packages/integration-tests/src/test-project-setup/data_storage_auth_with_triggers.ts#L171-L172 .
2. Add `TestProjectBase.setUp` (no-op in base class, override in this project) in addition to existing `tearDown` and hook it up into lifecycle

Reply:
my thinking here was originally 1, but if setup fails - that is not the same as deployment failure; so failure is in project creator. perhaps if we had a 'beforeEach' that called testProjectCreator.beforeEach or similar. I also wasn't sure if we were timing the deployment (haven't looked too far into this, didn't want that to be included in deploy time)- on that note - setup is almost instant/very fast. Opt 2. is good

Reply:
Some more context. The original idea behind those abstractions was :
`TestProjectCreator` - takes "project template" from `packages/integration-tests/src/test-projects` and renders it in temporary folder (perhaps it should be called `TestProjectCreatorOnDisk` or something like that.
Then classes implementing `TestProjectBase` are representing "rendered test project in temp directory" and provide operations for it.
I'd keep `TestProjectCreator` responsibility end at rendering files on disk and develop new hook in `TestProjectBase` for pre-deployment activities.